{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "#bokeh\n",
    "\n",
    "from bokeh.transform import linear_cmap\n",
    "from bokeh.models import ColorBar, ColumnDataSource\n",
    "from bokeh.palettes import Spectral6\n",
    "from bokeh.models import HoverTool\n",
    "from bokeh.core.properties import value\n",
    "from bokeh.io import output_file, show, save\n",
    "from bokeh.plotting import figure\n",
    "from IPython.display import display, HTML\n",
    "from bokeh.transform import factor_cmap, factor_mark\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#custom functions\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import generic_protein\n",
    "from skbio import Protein\n",
    "from pathlib import Path\n",
    "import os\n",
    "import logging\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import requests\n",
    "import wget\n",
    "\n",
    "\n",
    "\n",
    "class FastaMeta:\n",
    "    '''\n",
    "    object to hold fasta file and its corresponding meta file paths\n",
    "    '''\n",
    "    def __init__(self, fasta, meta):\n",
    "        '''\n",
    "        fasta: .fa file name\n",
    "        meta: .csv file name\n",
    "        '''\n",
    "        self.data_path = 'https://raw.githubusercontent.com/covid19-bh-machine-learning/master/master/data/'\n",
    "        self.fasta = os.path.join(self.data_path, fasta)\n",
    "        self.meta = os.path.join(self.data_path, meta)\n",
    "\n",
    "class DataProcessing(FastaMeta):\n",
    "    def __init__(self, fasta, meta):\n",
    "        super().__init__(fasta, meta)\n",
    "\n",
    "    def get_amino_df(self, k):\n",
    "        '''\n",
    "        k = kmer length\n",
    "        Generates all possible offsets of amino acid sequence and\n",
    "        returns a pandas dataframe merged with given metadata\n",
    "        meta_format ; csv tsv etc\n",
    "        '''\n",
    "        meta_df = pd.read_csv(self.meta, header=0)\n",
    "        seq_seq = defaultdict(list)\n",
    "        filename = wget.download(self.fasta)\n",
    "        seq_list = list(SeqIO.parse(filename, 'fasta', alphabet=generic_protein))\n",
    "        for s in seq_list:\n",
    "            for i in range(k):\n",
    "                seq_seq[f'seq_offset_{i}'].append(str(s.seq)[i:])\n",
    "        for key in seq_seq:\n",
    "            meta_df[key] = seq_seq[key]\n",
    "        return meta_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "orf1 = DataProcessing('coronavirus_orf1ab.fasta', 'coronavirus_orf1ab_meta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape WITH duplicates: (3046, 13)\n",
      "shape WITHOUT duplicates: (2384, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Length</th>\n",
       "      <th>Geo_Location</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>GenBank_Title</th>\n",
       "      <th>seq_offset_0</th>\n",
       "      <th>seq_offset_1</th>\n",
       "      <th>seq_offset_2</th>\n",
       "      <th>seq_offset_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YP_009555238</td>\n",
       "      <td>2019-02-21T00:00:00Z</td>\n",
       "      <td>Betacoronavirus 1</td>\n",
       "      <td>7095</td>\n",
       "      <td>USA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Orf1ab [Human coronavirus OC43]</td>\n",
       "      <td>MSKINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQK...</td>\n",
       "      <td>SKINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKL...</td>\n",
       "      <td>KINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKLE...</td>\n",
       "      <td>INKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKLET...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YP_002308478</td>\n",
       "      <td>2018-08-24T00:00:00Z</td>\n",
       "      <td>Bulbul coronavirus HKU11</td>\n",
       "      <td>6264</td>\n",
       "      <td>Hong Kong</td>\n",
       "      <td>Pycnonotus jocosus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-01</td>\n",
       "      <td>orf1ab polyprotein [Bulbul coronavirus HKU11-934]</td>\n",
       "      <td>MVKNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNL...</td>\n",
       "      <td>VKNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLV...</td>\n",
       "      <td>KNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLVT...</td>\n",
       "      <td>NVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLVTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YP_009513008</td>\n",
       "      <td>2018-08-24T00:00:00Z</td>\n",
       "      <td>Hedgehog coronavirus 1</td>\n",
       "      <td>7150</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Erinaceus europaeus</td>\n",
       "      <td>feces</td>\n",
       "      <td>2012</td>\n",
       "      <td>orf1ab [Betacoronavirus Erinaceus/VMC/DEU/2012]</td>\n",
       "      <td>MSSATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALS...</td>\n",
       "      <td>SSATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSP...</td>\n",
       "      <td>SATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSPW...</td>\n",
       "      <td>ATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSPWF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YP_009513020</td>\n",
       "      <td>2018-08-24T00:00:00Z</td>\n",
       "      <td>Coronavirus HKU15</td>\n",
       "      <td>6267</td>\n",
       "      <td>China: Hong Kong</td>\n",
       "      <td>Sus scrofa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010</td>\n",
       "      <td>replicase polyprotein [Porcine coronavirus HKU15]</td>\n",
       "      <td>MAKNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNL...</td>\n",
       "      <td>AKNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLY...</td>\n",
       "      <td>KNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLYA...</td>\n",
       "      <td>NKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLYAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>YP_009389424</td>\n",
       "      <td>2017-07-14T00:00:00Z</td>\n",
       "      <td>Wencheng Sm shrew coronavirus</td>\n",
       "      <td>6324</td>\n",
       "      <td>China</td>\n",
       "      <td>Suncus murinus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015</td>\n",
       "      <td>ORF1ab polyprotein [Wencheng Sm shrew coronavi...</td>\n",
       "      <td>MSVSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAY...</td>\n",
       "      <td>SVSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYG...</td>\n",
       "      <td>VSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYGL...</td>\n",
       "      <td>SKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYGLQ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Accession          Release_Date                        Species  Length  \\\n",
       "1  YP_009555238  2019-02-21T00:00:00Z              Betacoronavirus 1    7095   \n",
       "2  YP_002308478  2018-08-24T00:00:00Z       Bulbul coronavirus HKU11    6264   \n",
       "3  YP_009513008  2018-08-24T00:00:00Z         Hedgehog coronavirus 1    7150   \n",
       "4  YP_009513020  2018-08-24T00:00:00Z              Coronavirus HKU15    6267   \n",
       "5  YP_009389424  2017-07-14T00:00:00Z  Wencheng Sm shrew coronavirus    6324   \n",
       "\n",
       "       Geo_Location                 Host Isolation_Source Collection_Date  \\\n",
       "1               USA                  NaN              NaN             NaN   \n",
       "2         Hong Kong   Pycnonotus jocosus              NaN         2007-01   \n",
       "3           Germany  Erinaceus europaeus            feces            2012   \n",
       "4  China: Hong Kong           Sus scrofa              NaN            2010   \n",
       "5             China       Suncus murinus              NaN            2015   \n",
       "\n",
       "                                       GenBank_Title  \\\n",
       "1                    Orf1ab [Human coronavirus OC43]   \n",
       "2  orf1ab polyprotein [Bulbul coronavirus HKU11-934]   \n",
       "3    orf1ab [Betacoronavirus Erinaceus/VMC/DEU/2012]   \n",
       "4  replicase polyprotein [Porcine coronavirus HKU15]   \n",
       "5  ORF1ab polyprotein [Wencheng Sm shrew coronavi...   \n",
       "\n",
       "                                        seq_offset_0  \\\n",
       "1  MSKINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQK...   \n",
       "2  MVKNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNL...   \n",
       "3  MSSATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALS...   \n",
       "4  MAKNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNL...   \n",
       "5  MSVSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAY...   \n",
       "\n",
       "                                        seq_offset_1  \\\n",
       "1  SKINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKL...   \n",
       "2  VKNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLV...   \n",
       "3  SSATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSP...   \n",
       "4  AKNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLY...   \n",
       "5  SVSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYG...   \n",
       "\n",
       "                                        seq_offset_2  \\\n",
       "1  KINKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKLE...   \n",
       "2  KNVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLVT...   \n",
       "3  SATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSPW...   \n",
       "4  KNKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLYA...   \n",
       "5  VSKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYGL...   \n",
       "\n",
       "                                        seq_offset_3  \n",
       "1  INKYGLELHWAPEFPWMFEDAEEKLDNPSSSEVDMICSTTAQKLET...  \n",
       "2  NVSKRSPIVLPQIQPPPLQLFIAVAAAEEGHPKDLKYLGNYNLVTS...  \n",
       "3  ATGEGSQGARATYRAALNNEKRHDHVALTVPCCGTEAKVTALSPWF...  \n",
       "4  NKSKRDAIALPENVPPPLQLFIHVAAAEEGHPKVTTYLGNYNLYAT...  \n",
       "5  SKVELFVPISDEVDATHFGTFGDAVEAYASAAPSFEGVYFVAYGLQ...  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read for data folder and out put \n",
    "offset, kmer = 4, 4 # number of bases to offset\n",
    "amino_df = orf1.get_amino_df(offset)\n",
    "print(f\"shape WITH duplicates: {amino_df.shape}\")\n",
    "\n",
    "# remove duplicates\n",
    "amino_df.drop_duplicates(subset='Accession', keep=False, inplace=True)\n",
    "print(f\"shape WITHOUT duplicates: {amino_df.shape}\")\n",
    "amino_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check the hosts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\": \"JpP8FXbgAZLkfur7LiK3j9AGBhHNIvF742meBJrjO2ShJDhCG2I1uVvW+0DUtrmc\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\": \"xZlADit0Q04ISQEdKg2k3L4W9AwQBAuDs9nJL9fM/WwzL1tEU9VPNezOFX0nLEAz\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\": \"4BuPRZkdMKSnj3zoxiNrQ86XgNw0rYmBOxe7nshquXwwcauupgBF2DHLVG1WuZlV\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\": \"Dv1SQ87hmDqK6S5OhBf0bCuwAEvL5QYL0PuR/F1SPVhCS/r/abjkbpKDYL2zeM19\"};\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      if (url in hashes) {\n",
       "        element.crossOrigin = \"anonymous\";\n",
       "        element.integrity = \"sha384-\" + hashes[url];\n",
       "      }\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\"];\n",
       "  var css_urls = [];\n",
       "  \n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (var i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\": \"JpP8FXbgAZLkfur7LiK3j9AGBhHNIvF742meBJrjO2ShJDhCG2I1uVvW+0DUtrmc\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\": \"xZlADit0Q04ISQEdKg2k3L4W9AwQBAuDs9nJL9fM/WwzL1tEU9VPNezOFX0nLEAz\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\": \"4BuPRZkdMKSnj3zoxiNrQ86XgNw0rYmBOxe7nshquXwwcauupgBF2DHLVG1WuZlV\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\": \"Dv1SQ87hmDqK6S5OhBf0bCuwAEvL5QYL0PuR/F1SPVhCS/r/abjkbpKDYL2zeM19\"};\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      if (url in hashes) {\n        element.crossOrigin = \"anonymous\";\n        element.integrity = \"sha384-\" + hashes[url];\n      }\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\"];\n  var css_urls = [];\n  \n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (var i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of viral species:84\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from ipywidgets import interact\n",
    "import numpy as np\n",
    "\n",
    "from bokeh.io import push_notebook, show, output_notebook\n",
    "from bokeh.plotting import figure\n",
    "output_notebook()\n",
    "\n",
    "counts = Counter(amino_df['Species'])\n",
    "data = {'Species':list(counts.keys()), 'Count':list(counts.values())}\n",
    "source = ColumnDataSource(data=data)\n",
    "\n",
    "print(f\"Total number of viral species:{len(counts)}\")\n",
    "mapper = linear_cmap(field_name='Count', palette=Spectral6 ,low=min(data['Count']) ,high=max(data['Count']))\n",
    "\n",
    "p = figure(x_range=data['Species'], plot_height=400, plot_width=750, title=\"Virus Species\",\n",
    "           toolbar_location='right')\n",
    "pv = p.vbar(x='Species', top='Count', width=0.9, source=source, color=mapper)\n",
    "p.add_tools(HoverTool(\n",
    "#     renderers=[pv],\n",
    "    tooltips=[\n",
    "        ( 'Species',   '@'+'Species'            ), #state\n",
    "        ( 'Count',  '@'+'Count'            ), #count\n",
    "    ],\n",
    "\n",
    "    formatters={\n",
    "        \n",
    "    },\n",
    "\n",
    "    # display a tooltip whenever the cursor is vertically in line with a glyph\n",
    "    mode='vline'\n",
    "              ))\n",
    "\n",
    "p.xgrid.grid_line_color = None\n",
    "p.y_range.start = 0\n",
    "p.xaxis.visible = False\n",
    "p.yaxis.axis_label = 'Number of sequences'\n",
    "# output_file(\"bars.html\")\n",
    "# save(p)\n",
    "# display(HTML('bars.html'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def update(Species, Show_All):\n",
    "    if not Show_All:\n",
    "        c_ount = Counter(amino_df['Species'])\n",
    "        pv.data_source.data = {'Species':[Species], 'Count':[c_ount[Species]]}\n",
    "        print(\"> Check 'Show_All' to see the entire dataset.\")\n",
    "    else:\n",
    "        counts = Counter(amino_df['Species'])\n",
    "        pv.data_source.data = {'Species':list(counts.keys()), 'Count':list(counts.values())}\n",
    "        print(\"> Uncheck 'Show_All' to see the species level data.\")\n",
    "    push_notebook()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# An interactive bar chart of the coronavirus 'orf1ab' sequence counts in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"71449c40-8c2d-4c1b-879e-dfa062b46033\" data-root-id=\"1004\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"6303285e-5d42-426b-8546-f8d298ae2fda\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1015\"}],\"center\":[{\"id\":\"1017\"},{\"id\":\"1021\"}],\"left\":[{\"id\":\"1018\"}],\"plot_height\":400,\"plot_width\":750,\"renderers\":[{\"id\":\"1039\"}],\"title\":{\"id\":\"1005\"},\"toolbar\":{\"id\":\"1029\"},\"x_range\":{\"id\":\"1007\"},\"x_scale\":{\"id\":\"1011\"},\"y_range\":{\"id\":\"1009\"},\"y_scale\":{\"id\":\"1013\"}},\"id\":\"1004\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{},\"id\":\"1019\",\"type\":\"BasicTicker\"},{\"attributes\":{\"data\":{\"Count\":[219,3,4,97,1,2,4,2,2,3,2,2,60,2,4,2,520,2,3,9,5,2,2,2,2,2,2,74,16,2,2,327,5,6,4,8,1,9,10,10,39,60,239,516,5,1,1,1,1,1,1,1,1,1,39,2,1,1,1,1,1,1,1,1,1,1,1,1,1,2,1,1,1,1,1,10,1,1,4,1,1,1,1,1],\"Species\":[\"Betacoronavirus 1\",\"Bulbul coronavirus HKU11\",\"Hedgehog coronavirus 1\",\"Coronavirus HKU15\",\"Wencheng Sm shrew coronavirus\",\"Lucheng Rn rat coronavirus\",\"NL63-related bat coronavirus strain BtKYNL63-9b\",\"Rousettus bat coronavirus GCCDC1\",\"Ferret coronavirus\",\"Rhinolophus ferrumequinum alphacoronavirus HuB-2013\",\"BtRf-AlphaCoV/YN2012\",\"Nyctalus velutinus alphacoronavirus SC-2013\",\"Human coronavirus 229E\",\"Myotis ricketti alphacoronavirus Sax-2011\",\"China Rattus coronavirus HKU24\",\"Bat Hp-betacoronavirus Zhejiang2013\",\"Middle East respiratory syndrome-related coronavirus\",\"Mink coronavirus 1\",\"Bat coronavirus CDPHE15\",\"Bat coronavirus HKU10\",\"Rabbit coronavirus HKU14\",\"Common moorhen coronavirus HKU21\",\"White-eye coronavirus HKU16\",\"Wigeon coronavirus HKU20\",\"Night heron coronavirus HKU19\",\"Sparrow coronavirus HKU17\",\"Magpie-robin coronavirus HKU18\",\"Alphacoronavirus 1\",\"Murine coronavirus\",\"Thrush coronavirus HKU12\",\"Munia coronavirus HKU13\",\"Avian coronavirus\",\"Beluga whale coronavirus SW1\",\"Miniopterus bat coronavirus HKU8\",\"Bat coronavirus 1A\",\"Rhinolophus bat coronavirus HKU2\",\"Scotophilus bat coronavirus 512\",\"Pipistrellus bat coronavirus HKU5\",\"Tylonycteris bat coronavirus HKU4\",\"Rousettus bat coronavirus HKU9\",\"Human coronavirus HKU1\",\"Human coronavirus NL63\",\"Severe acute respiratory syndrome-related coronavirus\",\"Porcine epidemic diarrhea virus\",\"Pangolin coronavirus\",\"Tylonycteris pachypus bat coronavirus HKU4-related\",\"Miniopterus pusillus bat coronavirus HKU8-related\",\"Hipposideros pomona bat coronavirus CHB25\",\"Rousettus aegyptiacus bat coronavirus 229E-related\",\"Rhinolophus affinis bat coronavirus HKU2-related\",\"Hipposideros pomona bat coronavirus HKU10-related\",\"Scotophilus kuhlii bat coronavirus 512-related\",\"Miniopterus schreibersii bat coronavirus 1-related\",\"Pipistrellus abramus bat coronavirus HKU5-related\",\"Alphacoronavirus sp.\",\"Erinaceus hedgehog coronavirus HKU31\",\"Bat alphacoronavirus\",\"Swine enteric alphacoronavirus\",\"Coronavirus BtRs-AlphaCoV/YN2018\",\"Coronavirus BtSk-AlphaCoV/GX2018C\",\"Coronavirus BtRs-BetaCoV/YN2018D\",\"Coronavirus BtRt-BetaCoV/GX2018\",\"Coronavirus BtRs-BetaCoV/YN2018A\",\"Coronavirus BtSk-AlphaCoV/GX2018A\",\"Coronavirus BtSk-AlphaCoV/GX2018B\",\"Coronavirus BtRs-BetaCoV/YN2018B\",\"Coronavirus BtRs-BetaCoV/YN2018C\",\"Coronavirus BtSk-AlphaCoV/GX2018D\",\"Coronavirus BtRl-BetaCoV/SC2018\",\"Rhinolophus bat coronavirus HKU32\",\"Tylonycteris bat coronavirus HKU33\",\"Canada goose coronavirus\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/206679-3/2010\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/3398-19/2015\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/206645-41/2011\",\"Betacoronavirus sp.\",\"Feline alphacoronavirus 1\",\"Quail deltacoronavirus\",\"Sparrow deltacoronavirus\",\"Alphacoronavirus Mink/China/1/2016\",\"SARS-like coronavirus WIV16\",\"Camel coronavirus HKU23\",\"Alphacoronavirus 2\",\"Miniopterus bat coronavirus 1\"]},\"selected\":{\"id\":\"1048\"},\"selection_policy\":{\"id\":\"1049\"}},\"id\":\"1002\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"axis\":{\"id\":\"1018\"},\"dimension\":1,\"ticker\":null},\"id\":\"1021\",\"type\":\"Grid\"},{\"attributes\":{\"source\":{\"id\":\"1002\"}},\"id\":\"1040\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1047\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"1049\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"callback\":null,\"mode\":\"vline\",\"tooltips\":[[\"Species\",\"@Species\"],[\"Count\",\"@Count\"]]},\"id\":\"1041\",\"type\":\"HoverTool\"},{\"attributes\":{\"high\":520,\"low\":1,\"palette\":[\"#3288bd\",\"#99d594\",\"#e6f598\",\"#fee08b\",\"#fc8d59\",\"#d53e4f\"]},\"id\":\"1003\",\"type\":\"LinearColorMapper\"},{\"attributes\":{},\"id\":\"1022\",\"type\":\"PanTool\"},{\"attributes\":{},\"id\":\"1023\",\"type\":\"WheelZoomTool\"},{\"attributes\":{\"overlay\":{\"id\":\"1028\"}},\"id\":\"1024\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"1045\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{},\"id\":\"1025\",\"type\":\"SaveTool\"},{\"attributes\":{},\"id\":\"1026\",\"type\":\"ResetTool\"},{\"attributes\":{\"text\":\"Virus Species\"},\"id\":\"1005\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"1027\",\"type\":\"HelpTool\"},{\"attributes\":{\"factors\":[\"Betacoronavirus 1\",\"Bulbul coronavirus HKU11\",\"Hedgehog coronavirus 1\",\"Coronavirus HKU15\",\"Wencheng Sm shrew coronavirus\",\"Lucheng Rn rat coronavirus\",\"NL63-related bat coronavirus strain BtKYNL63-9b\",\"Rousettus bat coronavirus GCCDC1\",\"Ferret coronavirus\",\"Rhinolophus ferrumequinum alphacoronavirus HuB-2013\",\"BtRf-AlphaCoV/YN2012\",\"Nyctalus velutinus alphacoronavirus SC-2013\",\"Human coronavirus 229E\",\"Myotis ricketti alphacoronavirus Sax-2011\",\"China Rattus coronavirus HKU24\",\"Bat Hp-betacoronavirus Zhejiang2013\",\"Middle East respiratory syndrome-related coronavirus\",\"Mink coronavirus 1\",\"Bat coronavirus CDPHE15\",\"Bat coronavirus HKU10\",\"Rabbit coronavirus HKU14\",\"Common moorhen coronavirus HKU21\",\"White-eye coronavirus HKU16\",\"Wigeon coronavirus HKU20\",\"Night heron coronavirus HKU19\",\"Sparrow coronavirus HKU17\",\"Magpie-robin coronavirus HKU18\",\"Alphacoronavirus 1\",\"Murine coronavirus\",\"Thrush coronavirus HKU12\",\"Munia coronavirus HKU13\",\"Avian coronavirus\",\"Beluga whale coronavirus SW1\",\"Miniopterus bat coronavirus HKU8\",\"Bat coronavirus 1A\",\"Rhinolophus bat coronavirus HKU2\",\"Scotophilus bat coronavirus 512\",\"Pipistrellus bat coronavirus HKU5\",\"Tylonycteris bat coronavirus HKU4\",\"Rousettus bat coronavirus HKU9\",\"Human coronavirus HKU1\",\"Human coronavirus NL63\",\"Severe acute respiratory syndrome-related coronavirus\",\"Porcine epidemic diarrhea virus\",\"Pangolin coronavirus\",\"Tylonycteris pachypus bat coronavirus HKU4-related\",\"Miniopterus pusillus bat coronavirus HKU8-related\",\"Hipposideros pomona bat coronavirus CHB25\",\"Rousettus aegyptiacus bat coronavirus 229E-related\",\"Rhinolophus affinis bat coronavirus HKU2-related\",\"Hipposideros pomona bat coronavirus HKU10-related\",\"Scotophilus kuhlii bat coronavirus 512-related\",\"Miniopterus schreibersii bat coronavirus 1-related\",\"Pipistrellus abramus bat coronavirus HKU5-related\",\"Alphacoronavirus sp.\",\"Erinaceus hedgehog coronavirus HKU31\",\"Bat alphacoronavirus\",\"Swine enteric alphacoronavirus\",\"Coronavirus BtRs-AlphaCoV/YN2018\",\"Coronavirus BtSk-AlphaCoV/GX2018C\",\"Coronavirus BtRs-BetaCoV/YN2018D\",\"Coronavirus BtRt-BetaCoV/GX2018\",\"Coronavirus BtRs-BetaCoV/YN2018A\",\"Coronavirus BtSk-AlphaCoV/GX2018A\",\"Coronavirus BtSk-AlphaCoV/GX2018B\",\"Coronavirus BtRs-BetaCoV/YN2018B\",\"Coronavirus BtRs-BetaCoV/YN2018C\",\"Coronavirus BtSk-AlphaCoV/GX2018D\",\"Coronavirus BtRl-BetaCoV/SC2018\",\"Rhinolophus bat coronavirus HKU32\",\"Tylonycteris bat coronavirus HKU33\",\"Canada goose coronavirus\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/206679-3/2010\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/3398-19/2015\",\"Alphacoronavirus Bat-CoV/P.kuhlii/Italy/206645-41/2011\",\"Betacoronavirus sp.\",\"Feline alphacoronavirus 1\",\"Quail deltacoronavirus\",\"Sparrow deltacoronavirus\",\"Alphacoronavirus Mink/China/1/2016\",\"SARS-like coronavirus WIV16\",\"Camel coronavirus HKU23\",\"Alphacoronavirus 2\",\"Miniopterus bat coronavirus 1\"]},\"id\":\"1007\",\"type\":\"FactorRange\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1022\"},{\"id\":\"1023\"},{\"id\":\"1024\"},{\"id\":\"1025\"},{\"id\":\"1026\"},{\"id\":\"1027\"},{\"id\":\"1041\"}]},\"id\":\"1029\",\"type\":\"Toolbar\"},{\"attributes\":{},\"id\":\"1048\",\"type\":\"Selection\"},{\"attributes\":{\"start\":0},\"id\":\"1009\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"1011\",\"type\":\"CategoricalScale\"},{\"attributes\":{},\"id\":\"1013\",\"type\":\"LinearScale\"},{\"attributes\":{\"formatter\":{\"id\":\"1045\"},\"ticker\":{\"id\":\"1016\"},\"visible\":false},\"id\":\"1015\",\"type\":\"CategoricalAxis\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1028\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"1016\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"field\":\"Count\",\"transform\":{\"id\":\"1003\"}},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"field\":\"Count\",\"transform\":{\"id\":\"1003\"}},\"top\":{\"field\":\"Count\"},\"width\":{\"value\":0.9},\"x\":{\"field\":\"Species\"}},\"id\":\"1038\",\"type\":\"VBar\"},{\"attributes\":{\"fill_color\":{\"field\":\"Count\",\"transform\":{\"id\":\"1003\"}},\"line_color\":{\"field\":\"Count\",\"transform\":{\"id\":\"1003\"}},\"top\":{\"field\":\"Count\"},\"width\":{\"value\":0.9},\"x\":{\"field\":\"Species\"}},\"id\":\"1037\",\"type\":\"VBar\"},{\"attributes\":{\"axis\":{\"id\":\"1015\"},\"grid_line_color\":null,\"ticker\":null},\"id\":\"1017\",\"type\":\"Grid\"},{\"attributes\":{\"data_source\":{\"id\":\"1002\"},\"glyph\":{\"id\":\"1037\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1038\"},\"selection_glyph\":null,\"view\":{\"id\":\"1040\"}},\"id\":\"1039\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"axis_label\":\"Number of sequences\",\"formatter\":{\"id\":\"1047\"},\"ticker\":{\"id\":\"1019\"}},\"id\":\"1018\",\"type\":\"LinearAxis\"}],\"root_ids\":[\"1004\"]},\"title\":\"Bokeh Application\",\"version\":\"2.0.1\"}};\n",
       "  var render_items = [{\"docid\":\"6303285e-5d42-426b-8546-f8d298ae2fda\",\"notebook_comms_target\":\"1051\",\"root_ids\":[\"1004\"],\"roots\":{\"1004\":\"71449c40-8c2d-4c1b-879e-dfa062b46033\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1004"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "941646fdde54440f991a2c8ac7cff487",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Species', options=('Alphacoronavirus 1', 'Alphacoronavirus 2', 'Al…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# interact(update, f=amino_df['Species'].unique(), w=(0,50), A=(1,10), phi=(0, 20, 0.1))\n",
    "show(p, notebook_handle=True)\n",
    "fn = interact(update, Species=sorted(amino_df['Species'].unique()), Show_All=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
